### PR TITLE
Added Functionality to Inbox

### PR DESCRIPTION
**Pull request by paulmjojy**
The app will now fetch the emails corresponding to an email address from the server

### CODE REVIEW

File: TempMail/utils/Email.swift [hunk 1]
```diff
@@ -11,46 +11,73 @@ import SwiftUI
...
-    }
     
     // @TODO: implement real functionality
     // gets the contents of an individual message
     func getMessageContent(id: Int) -> MessageModel {
         return MessageModel(id: 1, from: "from@example.com", subject: "sample subject", date: "Today", attachments: nil, body: "somne text", textBody: "some text", htmlBody: "<h1> some text </h1>")
     }
 }
+
+struct Email_Previews: PreviewProvider {
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: Why is this necessary? Email is a class not a view

File: TempMail/utils/Email.swift [hunk 2]
```diff
@@ -11,46 +11,73 @@ import SwiftUI
...
         //list of acceptable chars
         let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
-        //      returns a random permutation of letters; string of up to length 9       for email addr
-        self.email_addr = String((0..<9).map{ _ in letters.randomElement()!})
+        //      returns a random permutation of letters; string of up to length 8 for email addr
+        self.email_addr = String((0..<8).map{ _ in letters.randomElement()!})
     }
     
+    // loads the Inbox that corresponds to the email address from the server, refreshing every second
+    func loadInbox() {
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: Every second is way too often. Every 30 sec or min is more reasonable. Don't forget that the user will have the option to pull to refresh.

File: TempMail/utils/Email.swift [hunk 3]
```diff
@@ -11,46 +11,73 @@ import SwiftUI
...
+        //creates the timer to refresh every second and run the code in the block
+        Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { timer in
+            //creates the url that will make the API call
+            guard let url = URL(string: "https://www.1secmail.com/api/v1/?action=getMessages&login=\(self.email_addr)&domain=1secmail.com")
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: The guard here doesn't make sense, URL won't verify if the url is valid or not. Also I think that this way of constructing the url is dirty. Use URLComponents() instead.

File: TempMail/utils/Email.swift [hunk 4]
```diff
@@ -11,46 +11,73 @@ import SwiftUI
...
+                else {
+                print("Invalid URL")
+                return
+            }
+            //creates the GET Request
+            let request = URLRequest(url: url)
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: Any reason why you're using URLRequest instead of a plain url?

File: TempMail/utils/Email.swift [hunk 5]
```diff
@@ -11,46 +11,73 @@ import SwiftUI
...
+            URLSession.shared.dataTask(with: request) { (data, response, error) in
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: I don't like the way you're handling errors here. You should explicitly check for errors instead of relying on a fall through. As it is, you get the same error if its a fetch error or if its a decode error which isn't helpful

File: TempMail/utils/Email.swift [hunk 6]
```diff
@@ -11,46 +11,73 @@ import SwiftUI
...
+                if let data = data {
+                    if let decodedResponse = try? JSONDecoder().decode([InboxModel].self, from: data) {
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: I think this decoding logic should be abstracted to a different function

File: TempMail/utils/DataModels.swift [hunk 7]
```diff
@@ -13,6 +13,13 @@ public struct InboxModel: Hashable, Codable, Identifiable {
     let from: String
     let subject: String
     let date: String
+    
+    enum CodingKeys: String, CodingKey {
```
Comment (aravindkoneru) [inline TempMail/utils/DataModels.swift]: What is this used for?

File: TempMail/ContentView.swift [hunk 8]
```diff
@@ -45,11 +45,11 @@ struct ContentView: View {
                 }.padding()
                 
                 
-                List(self.email.getInbox()) { message in
+                List(self.email.getEmails()) { message in
                     NavigationLink(destination: MessageView(message: self.email.getMessageContent(id: message.id))) {
                         InboxRow(messageInfo: message)
                     }
-                }
+                }.onAppear(perform: email.loadInbox)
```
Comment (aravindkoneru) [inline TempMail/ContentView.swift]: Why is this necessary? Wouldn't calling loadInbox() in the init() method of the Email class do the same thing?

**Review by aravindkoneru — COMMENTED** [review]
I'm gonna run some of this code on my computer and see if I understand everything correctly. At first glance, it doesn't look like this is following event driven programming

File: TempMail/utils/Email.swift (hunk 1 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: I never added that explicitly, it was being autogenerated. I'll remove it.

File: TempMail/utils/Email.swift (hunk 2 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: I know it says every second, but it is actually every 10 seconds when I tested it. I was under the impression that we were just going to let it refresh in the background. If we want to go with a pull to refresh then, I'll remove the timer. Is this what we want?

File: TempMail/utils/Email.swift (hunk 3 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: https://medium.com/@rbreve/displaying-a-list-with-swiftui-from-a-remote-json-file-6b4e4280a076

https://www.hackingwithswift.com/books/ios-swiftui/sending-and-receiving-codable-data-with-urlsession-and-swiftui

These are the 2 resources I was using to understand the concept of creating HTTP Get Requests and handling the JSON output. If I use URLComponents, I would immediately have to unwrap it to a URL anyway so that I can use it with the dataTask in line 39.

File: TempMail/utils/Email.swift (hunk 6 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: Right now the decoding logic is specific to the loading of the inbox. If, as I finish all the get requests for the individual message and attachment, I notice there is alot of similarity between the decoding of the 3 functions (which there may be) I will then abstract it to clean up the code.

File: TempMail/utils/Email.swift (hunk 3 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: However, I can remove the guard and the request and it still works, so I've went ahead and done that.

File: TempMail/utils/Email.swift (hunk 5 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: would you suggest a try catch? This is what I've changed it to:
                    if let data = data {
                        let decodedResponse = try JSONDecoder().decode([InboxModel].self, from: data)
                       ...
                    } else {
                        print("No Data to fetch")
                    }
                } catch {
                    print("Fetch failed: \(error.localizedDescription )")
                }

File: TempMail/utils/Email.swift (hunk 2 above)
Comment (aravindkoneru) [reply TempMail/utils/Email.swift]: It needs to be able to do both, think of the default mail app as an example of the desired functionality

Comment (paulmjojy) [reply TempMail/utils/Email.swift]: Ok, I will work on the pull to refresh after I finish the other http get requests.

Comment (paulmjojy) [reply TempMail/utils/Email.swift]: I am going to leave it at 1 sec for now so that I can test faster, but when it gets closer to production I'll set it to 30 sec - 1 min, and maybe have it loop in a different function.

File: TempMail/ContentView.swift [hunk 9]
```diff
@@ -45,7 +45,7 @@ struct ContentView: View {
                 }.padding()
                 
                 
-                List(self.email.getInbox()) { message in
+                List(self.email.getEmails()) { message in
```
Comment (aravindkoneru) [inline TempMail/ContentView.swift]: I think getInbox() is better than getEmails() since its more descriptive of the action being done (we could change the current getInbox -> fetchInbox)

File: TempMail/utils/Email.swift [hunk 10]
```diff
@@ -11,42 +11,62 @@ import SwiftUI
...
+    func loadInbox() -> Void {
+        //creates the timer to refresh every second and run the code in the block
+        Timer.scheduledTimer(withTimeInterval: 30, repeats: true) { timer in
+            //creates the url that will make the API call
+            let url = URL(string: "https://www.1secmail.com/api/v1/?action=getMessages&login=\(self.email_addr)&domain=1secmail.com")!
+            self.decodeInbox(url: url)
+        }
+    }
+    //sends request and decodes the JSON message that comes from server to get the inbox.
+    func decodeInbox(url: URL) -> Void {
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: This isn't what I meant when I said to abstract the JSON decode. I mean to say that we should abstract literally just the decoding (try JSONDecoder()....) because it would allow for better error handling. Check the playground for an example of what I meant

**Review by aravindkoneru — CHANGES_REQUESTED** [review]

File: TempMail/utils/Email.swift [hunk 11]
```diff
@@ -32,12 +32,16 @@ class Email : ObservableObject{
         let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
         //      returns a random permutation of letters; string of up to length 8 for email addr
         self.email_addr = String((0..<8).map{ _ in letters.randomElement()!})
+//        reset inbox
+        self.inbox = [InboxModel]()
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: This is bad because of side effects. Make a reset() function that's called when the button is pressed that sets a new email addr and sets the inbox to nil. Since we can use [InboxModel]? it makes more sense to do that instead of assigning it to be an empty array. We can also change the return type of setEmailAddr from a void to a String and set self.email_addr in the rest function

File: TempMail/utils/Email.swift [hunk 12]
```diff
@@ -55,8 +59,10 @@ class Email : ObservableObject{
                 guard let data = data else {return}
                 //updates instance variable
                 DispatchQueue.main.async {
-                    guard let inbox = self.parseInboxData(data: data) else {return}
-                    self.inbox = inbox
+                    if self.email_addr == email_addr {
+                        guard let inbox = self.parseInboxData(data: data) else {return}
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: Do this part async and not on the main thread. Don't want to block the main thread for some simple decoding that could've been done in the background. Since the decoding is going to be fast and resource light, it wouldn't matter if it ends up not being used.

**Review by aravindkoneru — CHANGES_REQUESTED** [review]

File: TempMail/ContentView.swift [hunk 13]
```diff
@@ -34,8 +34,7 @@ struct ContentView: View {
                 Spacer()
                 
                 Button(action: {
-                    self.email.setNewEmailAddr()
-                }) {
+                    self.email.reset()                }) {
```
Comment (aravindkoneru) [inline TempMail/ContentView.swift]: formatting

File: TempMail/utils/Email.swift [hunk 14]
```diff
@@ -81,12 +71,12 @@ class Email : ObservableObject{
...
     //  getter for email addr
     func getEmailAddr() -> String {
-        return email_addr
+        return email_addr!
     }
     
     //  getter for emails
     func getInbox() -> [InboxModel] {
-        return inbox
+        return inbox ?? [InboxModel]()
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: don't need this check here, just return [InboxModel]?

File: TempMail/utils/Email.swift [hunk 15]
```diff
@@ -10,31 +10,27 @@
...
+    func reset() -> Void {
+        self.inbox = nil
+        self.email_addr = setNewEmailAddr()
         loadInbox()
-        setNewEmailAddr()
     }
     
     //  generates an email address
-    func setNewEmailAddr() -> Void {
+    func setNewEmailAddr() -> String {
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: change to getNewEmailAddr()

File: TempMail/utils/Email.swift [hunk 16]
```diff
@@ -10,31 +10,27 @@
...
-        requestUrl.host = "1secmail.com"
-        requestUrl.path = "/api/v1/"
+        reset()
+    }
+    
+    //resets the email
+    func reset() -> Void {
+        self.inbox = nil
+        self.email_addr = setNewEmailAddr()
         loadInbox()
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: loadInbox() only needs to be called once, if its in reset, it'll be called multiple times. Move it to the init()

File: TempMail/utils/Email.swift [hunk 17]
```diff
@@ -10,31 +10,27 @@
 import SwiftUI
 
 class Email : ObservableObject{
-    @Published var email_addr: String
-    @Published var inbox: [InboxModel]
-    var requestUrl: URLComponents
+    @Published var email_addr: String?
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: Not the right idea conceptually since email_addr is a necessary property of the Email class. [InboxModel] is optional since an email doesn't need to have messages in the inbox. Change this to non-optional and initialize it in init()

File: TempMail/utils/Email.swift [hunk 18]
```diff
@@ -81,12 +71,12 @@ class Email : ObservableObject{
     
     //  getter for email addr
     func getEmailAddr() -> String {
-        return email_addr
+        return email_addr!
```
Comment (aravindkoneru) [inline TempMail/utils/Email.swift]: Change to just email_addr

**Review by aravindkoneru — CHANGES_REQUESTED** [review]

File: TempMail/utils/Email.swift (hunk 14 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: I believe you do need this check. Since the inbox is an optional, it can be nil. It is assigned nil until the user gets a mail. If I set it to return [InboxModel]?, then there will be an issue in line 49 of ContentView when we call getInbox, where the value of the optional type must be unwrapped. If nil is unwrapped, then there should be a default value, which I've set as an empty array. else if List(self.email.getInbox()) is force-unwrapped and the value is nil, the execution of the program will stop.

File: TempMail/utils/Email.swift (hunk 15 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: Why? I believe the purpose of this function is to set a new email address, not to fetch it. getEmailAddr() and getNewEmailAddr might be too similar in name as well, which is another issue in my opinion. If you don't believe this is a set method, I could change the func signature to generateEmailAddr, or generateNewEmailAddr. What do you think?

File: TempMail/utils/Email.swift (hunk 14 above)
Comment (aravindkoneru) [reply TempMail/utils/Email.swift]: The check for nil needs to be changed on the gui side, I'll make the change. As it is, there's no functional difference between nil and [], they're basically placeholders for each other. Conceptually if an inbox has no messages (ie []) then it is nil and vice versa. Since we have optionals, we should use them as intended instead of reverting to familiar behavior. It would be nicer to not show the list of messages if there are no emails in your inbox similar to how the mail app does it. The gui would check if the value is nil then show nothing else show the list of messages.

File: TempMail/utils/Email.swift (hunk 15 above)
Comment (aravindkoneru) [reply TempMail/utils/Email.swift]: Good point, genEmailAddr is a better name

File: TempMail/utils/Email.swift (hunk 14 above)
Comment (paulmjojy) [reply TempMail/utils/Email.swift]: Ok that sounds good. Maybe the gui can check for nil, and if nil, show a grayed out background message along the lines of "No emails currently" or "Mailbox is empty"

File: TempMail/utils/Email.swift (hunk 15 above)
Comment (aravindkoneru) [reply TempMail/utils/Email.swift]: Actually, genNewEmailAddr is better, its more explicit

**Review by aravindkoneru — APPROVED** [review]
